### PR TITLE
refactor(pipeline): simplify OCI SkyPilot workflow + config

### LINKED ISSUE #791: refactor(pipeline): simplify OCI SkyPilot workflow + config
## Why

The OCI SkyPilot integration landed alongside RunPod as a second backend, but the surface area is tangled with duplication, multi-shape pinning, and deeply-quoted CI heredocs. Cleanup is overdue before more compute providers are added.

## Smells being addressed

- Inlined worker-bootstrap block in both `runpod-template.yaml` + `oci-cpu-template.yaml` duplicates `scripts/skypilot_worker_bootstrap.sh`. Stale chicken-and-egg comment ("isn't on the not-yet-rebuilt dev-snapshot image until the checkout itself lands") no longer applies — script is on `main`, image has rebuilt.
- Two different shapes for "what image to pull" (`image_id: docker:…` for RunPod vs custom `envs.WORKER_IMAGE` for OCI), with a per-provider `pin_search`/`pin_replace` sed step in CI to swap the tag.
- 70-line `docker run … bash -c '…'` in `test-dataset-generation.yml` with three levels of quoting carrying provider switch + cred-write + bridge-install + sky check + launcher invocation.
- `configs/compute/oci-debug-template.yaml` self-documents as not wired into any matrix — dead.
- OCI cell still `continue-on-error: true` from initial bedding-in.
- Stale launcher docstring referencing pre-merge phasing.

## Scope

Covered in PR (linked below):

- Both YAMLs call `bash scripts/skypilot_worker_bootstrap.sh` instead of inline checkout.
- Launcher gains `--worker-image-tag`; `_override_image_id` injects `docker:<image>` for non-OCI backends, `WORKER_IMAGE` env always injected for OCI's sub-docker. Sed-pin step deleted.
- Per-provider cred-write extracted to `scripts/skypilot_write_provider_creds.sh`.
- `skypilot[oci]` confirmed pinned in `requirements-app.txt`. Runtime `pip install` bridge kept this PR (one-line follow-up removes after dev-snapshot rebuild).
- `oci-debug-template.yaml` deleted; references in `docs/doc-map.yaml` + `test-skypilot-debug.yml` cleaned up.
- OCI flipped to blocking; per-job `continue_on_error` matrix indirection collapsed.
- Stale comments trimmed.

## Out of scope

- The `--privileged + --ulimit nofile` hammer on the OCI `docker run` is tracked under #776 — not touched here.
- Removing the runtime `pip install "skypilot[oci]"` bridge is a one-line follow-up after this PR's `requirements-app.txt` triggers a dev-snapshot rebuild.

## Acceptance

- Net reduction in lines across the integration.
- Both `runpod` and `oci` cells of `test-dataset-generation.yml` green.
- `make format` + `make test` clean.
- No remaining references to `oci-debug-template.yaml`, `pin_search`, `pin_replace`, or `WORKER_IMAGE.*dev-snapshot` literal.

### PR DESCRIPTION
**Pull request by ktinubu**
## Summary

- Both compute templates now call `bash scripts/skypilot_worker_bootstrap.sh` instead of duplicating an inline `WORKER_GIT_REF` checkout block. Stale chicken-and-egg comments removed.
- `pipeline.entrypoints.skypilot_launch_smoke` gains `--worker-image-tag`. New `_override_image_id` helper iterates `task.resources`, skips OCI via `isinstance(res.cloud, SkyOCI)`, rebuilds non-OCI resources with `res.copy(image_id=f"docker:{worker_image}")`. `WORKER_IMAGE` env always injected for OCI's sub-docker. Workflow's per-provider `pin_search`/`pin_replace` sed step deleted.
- Per-provider cred-write extracted to `scripts/skypilot_write_provider_creds.sh` (Google Shell Style: `#!/usr/bin/env bash`, `set -euo pipefail`, `[[ ]]`, quoted expansions). Workflow heredoc shrunk dramatically.
- `skypilot[oci]` confirmed pinned in `requirements-app.txt` as `skypilot[runpod,oci]==0.12.0`. Runtime `pip install` bridge kept this PR — one-line follow-up removes it after dev-snapshot rebuilds.
- `configs/compute/oci-debug-template.yaml` deleted (was not wired into any matrix). References in `docs/doc-map.yaml` and `.github/workflows/test-skypilot-debug.yml` cleaned up.
- OCI flipped to blocking: `continue_on_error: true` → `false` in all three matrix sections, then the per-job `continue-on-error` indirection collapsed entirely.
- Stale launcher docstring + OCI YAML header comments trimmed.

Net diff: −47 lines across 8 files.

## Out of scope

- The `--privileged + --ulimit nofile` capset on the OCI `docker run` — tracked under #776.
- Removing the runtime `pip install "skypilot[oci]"` bridge — one-line follow-up after dev-snapshot rebuilds with this PR's `requirements-app.txt`.

## Test plan

- [x] `make format` clean (pyright, ruff, shellcheck, all pre-commit hooks)
- [x] `make test` — 296 passed, including all 25 launcher tests
- [x] `bash -n scripts/skypilot_write_provider_creds.sh` — syntax OK
- [x] Both compute templates parse via `yaml.safe_load`
- [x] `python -m pipeline.entrypoints.skypilot_launch_smoke --help` shows `--worker-image-tag`
- [x] Smoke check on `_override_image_id`: RunPod template → `image_id={None: 'docker:tinaudio/synth-setter:test123'}`; OCI template → unchanged
- [x] No remaining references to `oci-debug-template`, `pin_search`, `pin_replace`, or `WORKER_IMAGE.*dev-snapshot` literal
- [ ] CI matrix: `oci` cell of `test-dataset-generation.yml` green on this PR (same-repo PR events run OCI-only by policy; RunPod coverage exercised via `workflow_dispatch` with `provider=runpod` or `provider=all`)
- [ ] `validate-shard` job downloads + validates every shard from R2

Closes #791

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

Behavioral checks for each promise in the PR. The Level 1 end-to-end check (CI matrix) is in flight and will be updated when complete.

- [x] **Launcher injects `image_id` for non-OCI backends and `WORKER_IMAGE` for OCI** *(Level 2 — invoked `_override_image_id` directly on both production templates; promise is per-backend mutation logic, not end-to-end)*
  ```bash
  $ python3 -c "
  import sky
  from pipeline.entrypoints.skypilot_launch_smoke import _override_image_id
  task_rp = sky.Task.from_yaml('configs/compute/runpod-template.yaml')
  print('RunPod BEFORE:', [r.image_id for r in task_rp.resources])
  _override_image_id(task_rp, 'tinaudio/synth-setter:test-tag-abc123')
  print('RunPod AFTER: ', [r.image_id for r in task_rp.resources])
  task_oci = sky.Task.from_yaml('configs/compute/oci-cpu-template.yaml')
  print('OCI BEFORE:   ', [r.image_id for r in task_oci.resources])
  _override_image_id(task_oci, 'tinaudio/synth-setter:test-tag-abc123')
  print('OCI AFTER:    ', [r.image_id for r in task_oci.resources])
  "
  RunPod BEFORE: [None, None, None, None, None, None, None]
  RunPod AFTER:  [{None: 'docker:tinaudio/synth-setter:test-tag-abc123'}, ...×7]
  OCI BEFORE:    [None]
  OCI AFTER:     [None]
  ```
  RunPod's 7-resource alternative set all get the new image_id; OCI's resource is left untouched (correct — OCI runs the worker via the YAML's inline `docker run`, fed by `WORKER_IMAGE` env).

- [x] **`scripts/skypilot_write_provider_creds.sh` writes correct files with mode 0600 for both providers** *(Level 2 — invoked the script with fake env vars in a sandboxed `$HOME`; promise is "writes the right files, with the right permissions, with the right content")*
  ```bash
  $ TMPHOME=$(mktemp -d); HOME="$TMPHOME" PROVIDER=runpod RUNPOD_API_KEY=fake-key \
      bash scripts/skypilot_write_provider_creds.sh
  $ stat -c '%a %n' "$TMPHOME/.runpod/config.toml"
  600 /tmp/tmp.XXXX/.runpod/config.toml
  $ cat "$TMPHOME/.runpod/config.toml"
  [default]
  api_key = "fake-key"

  $ TMPHOME=$(mktemp -d); HOME="$TMPHOME" PROVIDER=oci OCI_USER_OCID=… \
      OCI_TENANCY_OCID=… OCI_FINGERPRINT=… OCI_REGION=us-ashburn-1 \
      OCI_COMPARTMENT_OCID=… OCI_API_KEY_PEM='…' OCI_IMAGE_TAG=skypilot:cpu-ubuntu-2204 \
      bash scripts/skypilot_write_provider_creds.sh
  $ stat -c '%a %n' "$TMPHOME/.oci/config" "$TMPHOME/.oci/oci_api_key.pem" "$TMPHOME/.sky/config.yaml"
  600 /tmp/tmp.YYYY/.oci/config
  600 /tmp/tmp.YYYY/.oci/oci_api_key.pem
  600 /tmp/tmp.YYYY/.sky/config.yaml
  ```
  All three OCI files at mode 600 (umask 077 honored); both `~/.oci/config` and `~/.sky/config.yaml` contain the expected fields including `compartment_ocid` and `image_tag_general`.

- [x] **`--worker-image-tag` is exposed by the launcher CLI** *(Level 2 — invoked `--help` on the actual entrypoint; promise is API surface)*
  ```bash
  $ python3 -m pipeline.entrypoints.skypilot_launch_smoke --help 2>&1 | grep -A2 -- "--worker-image-tag"
    --worker-image-tag TEXT  Worker Docker image tag (under tinaudio/synth-
                             setter). Injected as WORKER_IMAGE env for the OCI
                             sub-docker invocation, and as Resources.image_id
  ```

- [x] **Both production templates invoke the shared bootstrap script (no inline checkout block)** *(Level 4 — query the file contents, since the YAML's *promise* is "calls bootstrap" — Level 1 would over-specify this and require provisioning compute)*
  ```bash
  $ grep -n "skypilot_worker_bootstrap.sh" configs/compute/runpod-template.yaml configs/compute/oci-cpu-template.yaml
  configs/compute/runpod-template.yaml:57:  bash /home/build/synth-setter/scripts/skypilot_worker_bootstrap.sh
  configs/compute/oci-cpu-template.yaml:66:    bash scripts/skypilot_worker_bootstrap.sh
  ```
  The only remaining `WORKER_GIT_REF` reference in the templates is the OCI docker-run env-forward (`-e WORKER_GIT_REF \`) — required so the script inside the OCI container can read it. The inline checkout block is gone in both files.

- [x] **No stale references to deleted/renamed elements** *(Level 4 — repo-wide grep)*
  ```bash
  $ grep -rn "oci-debug-template" .github/ configs/ docs/ scripts/ pipeline/ src/
  (no output)
  $ grep -rn "pin_search\|pin_replace" .github/
  (no output)
  $ grep -n 'WORKER_IMAGE.*dev-snapshot\|image_id: docker:tinaudio.*dev-snapshot' \
      configs/compute/runpod-template.yaml configs/compute/oci-cpu-template.yaml
  (no output — production templates rely on launcher injection)
  ```
  The 8 RunPod *debug* templates still carry their `dev-snapshot` literals — those are out of scope (used by `test-skypilot-debug.yml`'s diagnostic matrix, not the production launcher).

- [x] **`skypilot[oci]` SDK is pinned (no longer relying on the runtime bridge)** *(Level 5 — file content; Level 2 invocation `python -c "import oci"` would only test the local venv, not the dev-snapshot image. The actual image-import check is the Level 1 CI matrix below.)*
  ```bash
  $ grep -E '^skypilot' requirements*.txt
  requirements-app.txt:skypilot[runpod,oci]==0.12.0
  ```
  Per the PR description, the runtime `pip install "skypilot[oci]==0.12.0"` bridge stays this PR as belt-and-braces; one-line follow-up removes it after dev-snapshot rebuilds.

- [x] **All three SkyPilot shell scripts have valid bash syntax** *(Level 2 — `bash -n` parses each file)*
  ```bash
  $ bash -n scripts/skypilot_write_provider_creds.sh && echo OK
  OK
  $ bash -n scripts/skypilot_worker_bootstrap.sh && echo OK
  OK
  $ bash -n scripts/skypilot_worker_run.sh && echo OK
  OK
  ```

- [ ] **End-to-end CI matrix: `runpod` and `oci` cells of `test-dataset-generation.yml` both green** *(Level 1 — the actual end-to-end promise; auto-triggered by this PR open. Will update when CI completes.)*
  ```bash
  $ gh pr checks 792 --repo tinaudio/synth-setter | grep -E "Launch generate_dataset"
  Launch generate_dataset on runpod via SkyPilot   pending  0  https://github.com/tinaudio/synth-setter/actions/runs/25326941181/job/74250137438
  Launch generate_dataset on oci via SkyPilot      pending  0  https://github.com/tinaudio/synth-setter/actions/runs/25326941181/job/74250137592
  ```

### Adversarial / implicit-contract checks

- [x] **Cred-write script fails loudly on missing env vars** *(Level 2 — invoked with `RUNPOD_API_KEY` deliberately unset; promise is "no silent partial config files left on disk")*
  ```bash
  $ HOME=$(mktemp -d) PROVIDER=runpod bash scripts/skypilot_write_provider_creds.sh; echo "rc=$?"
  ::error::RUNPOD_API_KEY is empty
  rc=1
  ```
  Script uses GHA's `::error::` annotation format; nonzero exit means downstream steps will short-circuit before reaching `sky check`.

- [x] **OCI template still parses through SkyPilot's loader after the bootstrap-script swap** *(Level 2 — confirmed by the `_override_image_id` probe above, which calls `sky.Task.from_yaml('configs/compute/oci-cpu-template.yaml')` successfully)*

- [x] **Removing `image_id` literal from `runpod-template.yaml` doesn't break `from_yaml` parse** *(Level 2 — same probe confirmed `task_rp = sky.Task.from_yaml('configs/compute/runpod-template.yaml')` succeeds, returning resources with `image_id=None` until the launcher injects)*

File: pipeline/entrypoints/skypilot_launch_smoke.py [hunk 1]
```diff
@@ -8,8 +8,6 @@
 `--num-workers N>1` fans out N single-node clusters in parallel (RunPod doesn't
 support num_nodes>1). Each rank gets SYNTH_SETTER_WORKER_RANK /
 SYNTH_SETTER_NUM_WORKERS injected; one shared spec → one r2_prefix.
-
-First scaffolding under #534. No `compute_config` schema yet — Phase A–C PRs.
 """
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply pipeline/entrypoints/skypilot_launch_smoke.py]: Fixed in ec26c3b — module docstring rewritten as provider-neutral and now documents per-backend image handling (RunPod `image_id: docker:<image>` vs OCI sub-docker via `WORKER_IMAGE` env).

File: pipeline/entrypoints/skypilot_launch_smoke.py [hunk 2]
```diff
@@ -23,6 +21,7 @@
 import click
 import sky
 from dotenv import dotenv_values
+from sky.clouds import OCI as SkyOCI
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply pipeline/entrypoints/skypilot_launch_smoke.py]: Fixed in ec26c3b — moved the OCI cloud-class import inside `_override_image_id` with an `ImportError` fallback. When `skypilot[oci]` extras aren't installed (the RunPod CI cell on dev-snapshot images), `oci_cls` is `None` and the isinstance check is skipped — every resource is treated as non-OCI, which is correct for that image.

[the remark this answers is not in the document]
Comment (ktinubu) [reply pipeline/entrypoints/skypilot_launch_smoke.py]: Fixed in ec26c3b — moved the OCI cloud-class import inside `_override_image_id` with an `ImportError` fallback. When `skypilot[oci]` extras aren't installed (the RunPod CI cell on dev-snapshot images), `oci_cls` is `None` and the isinstance check is skipped — every resource is treated as non-OCI, which is correct for that image.

File: pipeline/entrypoints/skypilot_launch_smoke.py [hunk 3]
```diff
@@ -189,13 +190,26 @@ def upload_spec_to_r2(spec: DatasetPipelineSpec, cluster_name: str) -> str:
...
     config_path: Path,
     template_path: Path,
     env_file_path: Path,
     cluster_name: str | None,
     spec_out: Path | None,
     num_workers: int,
+    worker_image_tag: str,
 ) -> None:
     """Launch the smoke `generate_dataset` run on RunPod via SkyPilot."""
     if num_workers < 1:
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply pipeline/entrypoints/skypilot_launch_smoke.py:227]: Fixed in ec26c3b — `main()` docstring now reads "Launch the smoke `generate_dataset` run via SkyPilot (RunPod or OCI per `--template`)."

File: configs/compute/oci-cpu-template.yaml [hunk 4]
```diff
@@ -1,35 +1,27 @@
...
   RCLONE_CONFIG_R2_PROVIDER: ""
   RCLONE_CONFIG_R2_ACCESS_KEY_ID: ""
   RCLONE_CONFIG_R2_SECRET_ACCESS_KEY: ""
   RCLONE_CONFIG_R2_ENDPOINT: ""
   WORKER_SPEC_URI: ""
-  # Default below is the CI sed pin target — see "Pin worker image tag"
-  # step in .github/workflows/test-dataset-generation.yml.
-  WORKER_IMAGE: "tinaudio/synth-setter:dev-snapshot"
+  WORKER_IMAGE: ""
   WORKER_GIT_REF: ""
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply configs/compute/oci-cpu-template.yaml:25]: We should fail loudly if launcher env injection fails

File: configs/compute/runpod-template.yaml [hunk 5]
```diff
@@ -12,8 +12,7 @@
 #
 # Image: tinaudio/synth-setter:dev-snapshot must be pushed to Docker Hub
 # (`make docker-build-dev-snapshot` then `docker push`) before launch.
-# Tag pinning to a git SHA is deferred to the Phase B PR per the design doc
-# at docs/design/skypilot-compute-integration.md.
+# `image_id` is set per-launch by the launcher from --worker-image-tag.
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply configs/compute/runpod-template.yaml]: Fixed in 40ed8de — runpod-template.yaml header now reads `tinaudio/synth-setter:<worker-image-tag>` with `dev-snapshot` called out as the default tag, so it tracks the launcher CLI surface instead of pinning operators to one tag.

File: configs/compute/runpod-template.yaml [hunk 6]
```diff
@@ -12,8 +12,7 @@
...
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply configs/compute/runpod-template.yaml]: Left as-is — same stance as the OCI sibling thread (#3182574743): if the launcher fails to inject `image_id`, we want the YAML to fail loudly at provision time rather than silently fall back to a stale default that masks the bug. The runpod-template is launcher-driven; standalone `sky launch` of it is not a supported path.

File: configs/compute/oci-cpu-template.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ktinubu) [reply configs/compute/oci-cpu-template.yaml:25]: Fixed in eab2b60 — both `setup:` and `run:` blocks now bail with `: "${WORKER_IMAGE:?WORKER_IMAGE must be set by the launcher (--worker-image-tag); refusing to docker pull/run an unset image}"` before any `docker pull`/`docker run`, so launcher env-injection failure now produces a clear error instead of `docker pull ""`.

File: .github/workflows/test-dataset-generation.yml [hunk 7]
```diff
@@ -263,27 +264,17 @@ jobs:
...
-    continue-on-error: ${{ matrix.continue_on_error }}
     strategy:
       fail-fast: false
       matrix:
-        include:
-          - provider: runpod
-            continue_on_error: false
-          - provider: oci
-            continue_on_error: true
+        provider: ${{ fromJSON(needs.setup.outputs.validate_matrix) }}
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply .github/workflows/test-dataset-generation.yml:282]: Already covered in current code — the validate jobs use `if: needs.setup.outputs.has_jobs == 'true' && !cancelled()`, not `success()`. `!cancelled()` evaluates true when `generate` partially fails, so `validate-spec` and `validate-shard` still run. Combined with `fail-fast: false` on the generate matrix and on the validate matrices, a RunPod transient does not cancel sibling cells, and OCI validation proceeds independently. Each per-provider validate cell downloads its own provider-named artifact (`test-run-metadata-${{ matrix.provider }}`), so a missing artifact only fails the cell whose generate cell crashed before upload — `if: always()` on the generate upload step keeps that rare too. No code change needed.

File: docs/reference/configuration-reference.md [hunk 8]
```diff
@@ -124,13 +124,14 @@ The RunPod template exists today (data-pipeline smoke). Vast.ai template not yet
 shape (see the file for the full template, including the inline `os._exit(0)`
 workaround for [#735](https://github.com/tinaudio/synth-setter/issues/735)):
 
+The launcher injects `image_id` per-launch via `--worker-image-tag` (default `dev-snapshot`) for non-OCI backends, so the template no longer carries the literal:
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply docs/reference/configuration-reference.md]: Fixed in eab2b60 — prose now explicitly names the removed value: "the template no longer carries a literal `image_id: docker:tinaudio/synth-setter:dev-snapshot`".

File: docs/reference/github-actions.md [hunk 9]
```diff
@@ -24,14 +24,14 @@ For GitHub Actions concepts, see [GitHub's docs](https://docs.github.com/en/acti
...
+| `test-spec-materialization` | Exercises `spec-materialization` and validates test-specific config values.                                                                                                                                                              |                                                                                                                                              |
+| `flush-investigation`       | Runs the `flush-investigation.ipynb` notebook in Docker, uploads rendered HTML + audio as an artifact (90-day retention).                                                                                                                |                                                                                                                                              |
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply docs/reference/github-actions.md]: Fixed in eab2b60 — the `test-dataset-generation` row now reads: "Dynamic matrix… Same-repo PR events run **OCI-only**; fork PRs run an empty matrix (no secrets); `workflow_dispatch` selects via `inputs.provider` (`all` / `runpod` / `oci`)."

File: .github/workflows/test-dataset-generation.yml [hunk 10]
```diff
@@ -55,45 +54,94 @@ concurrency:
...
+            if [[ "${PR_HEAD_REPO}" == "${REPO}" ]]; then
+              providers=("oci")
+            fi
+          elif [[ "${EVENT_NAME}" == "workflow_dispatch" ]]; then
+            case "${PROVIDER_INPUT}" in
+              runpod) providers=("runpod") ;;
+              oci)    providers=("oci") ;;
+              *)      providers=("runpod" "oci") ;;
+            esac
+          fi
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply .github/workflows/test-dataset-generation.yml:98]: Fixed by updating the PR test plan — the checkbox now reads "`oci` cell of `test-dataset-generation.yml` green on this PR (same-repo PR events run OCI-only by policy; RunPod coverage exercised via `workflow_dispatch` with `provider=runpod` or `provider=all`)." The `setup` job's policy is intentional (RunPod is more flake-prone and has been moved to dispatch-only for PR runs); the test plan was the stale half.

File: .github/workflows/test-dataset-generation.yml [hunk 11]
```diff
@@ -24,7 +23,7 @@ on:
         required: false
         default: "dev-snapshot"
       provider:
-        description: "Provider(s) to launch on (manual runs only; PR runs always exercise all)"
+        description: "Provider(s) to launch on (manual runs only; PR runs are OCI-only)"
         required: false
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply .github/workflows/test-dataset-generation.yml:30]: Fixed in eab2b60 — the workflow header now reads: "The provider matrix is computed at runtime by the `setup` job (see its `Policy:` block): same-repo PR → OCI only; fork PR → empty (no secrets); workflow_dispatch → respects `inputs.provider` (all/runpod/oci)."

File: pipeline/entrypoints/skypilot_launch_smoke.py [hunk 12]
```diff
@@ -189,15 +198,28 @@ def upload_spec_to_r2(spec: DatasetPipelineSpec, cluster_name: str) -> str:
...
+    "--worker-image-tag",
+    type=str,
+    default="dev-snapshot",
+    show_default=True,
+    help=(
+        "Worker Docker image tag (under tinaudio/synth-setter). Injected as WORKER_IMAGE env "
+        "for the OCI sub-docker invocation, and as Resources.image_id for backends that accept "
+        "`docker:<image>` (e.g. RunPod). OCI's backend rejects `docker:<image>` so its "
+        "image_id is left untouched."
+    ),
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply pipeline/entrypoints/skypilot_launch_smoke.py:215]: Fixed in eab2b60 — `--worker-image-tag` is now validated against the OCI distribution tag grammar (`^[A-Za-z0-9_][A-Za-z0-9_.-]{0,127}$`) at the top of `main()`, raising a `ClickException` with a clear message before any docker ref is constructed. New parameterized regression test covers six invalid-tag inputs (`/`, `:`, whitespace, leading dot/dash, etc.) plus boundary cases.

File: pipeline/entrypoints/skypilot_launch_smoke.py [hunk 13]
```diff
@@ -256,10 +289,42 @@ def main(
         )
 
 
+def _override_image_id(task: sky.Task, worker_image: str) -> None:
+    """Pin every Resources entry's image_id to ``docker:<worker_image>`` for backends that take it.
+
+    SkyPilot's OCI backend rejects ``image_id: docker:<image>`` — that path runs the worker via
+    a sub-docker invocation inside the YAML's run: block and consumes WORKER_IMAGE from env, so
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply pipeline/entrypoints/skypilot_launch_smoke.py:296]: Fixed in 4b2d721 — added `TestOverrideImageId` (4 cases): (1) non-OCI Resource gets `image_id` rewritten to `docker:<worker_image>`, (2) every entry in a multi-Resources alt-set is mutated (RunPod ships 7), (3) OCI Resources are passed through untouched and `set_resources` is not called when nothing mutated, (4) when `skypilot[oci]` extras are missing the lazy `from sky.clouds import OCI` raises `ImportError` and the function falls back to treating every Resource as non-OCI without crashing. All four pass locally.

File: .github/workflows/test-dataset-generation.yml [hunk 14]
```diff
@@ -199,40 +224,19 @@ jobs:
...
-                *)
-                  echo "ERROR: unknown PROVIDER=$PROVIDER" >&2
-                  exit 1
-                  ;;
-              esac
+              bash scripts/skypilot_write_provider_creds.sh
+              if [[ "$PROVIDER" == "oci" ]]; then
+                if ! python -c "import oci" 2>/dev/null; then
+                  echo "[oci-bridge] dev-snapshot lacks oci sdk; installing skypilot[oci] at runtime"
+                  pip install --quiet "skypilot[oci]==0.12.0" || { echo "::error::oci bridge install failed" >&2; exit 1; }
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply .github/workflows/test-dataset-generation.yml:233]: Fixed in 4b2d721 — `on.pull_request.paths` now includes both shared launcher scripts:

```yaml
- "scripts/skypilot_worker_run.sh"
- "scripts/skypilot_worker_bootstrap.sh"
- "scripts/skypilot_write_provider_creds.sh"
- "pipeline/entrypoints/skypilot_launch_smoke.py"
```

so future edits to `skypilot_worker_bootstrap.sh` (used by both `runpod-template.yaml` and `oci-cpu-template.yaml`) and `skypilot_write_provider_creds.sh` (used by the workflow itself) will trigger this smoke workflow.
